ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc

```diff
@@ -0,0 +1,225 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * GStreamer / NNStreamer gRPC/protobuf support
+ * Copyright (C) 2020 Dongju Chae <dongju.chae@samsung.com>
+ */
+/**
+ * @file    nnstreamer_grpc_protobuf.cc
+ * @date    21 Oct 2020
+ * @brief   gRPC/Protobuf wrappers for nnstreamer
+ * @see     https://github.com/nnstreamer/nnstreamer
+ * @author  Dongju Chae <dongju.chae@samsung.com>
+ * @bug     No known bugs except for NYI items
+ */
+
+#include "nnstreamer_grpc_protobuf.h"
+
+#include <nnstreamer_log.h>
+#include <nnstreamer_plugin_api.h>
+
+#include <thread>
+
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+
+#include <gst/base/gstdataqueue.h>
+
+using nnstreamer::protobuf::TensorService;
+using nnstreamer::protobuf::Tensors;
+using nnstreamer::protobuf::Tensor;
+
+using google::protobuf::Empty;
+
+using namespace grpc;
+
+/** @brief constructor */
+ServiceImplProtobuf::ServiceImplProtobuf (gboolean is_server,
+    const gchar *host, const gint port):
+  NNStreamerRPC (is_server, host, port), client_stub_ (nullptr)
+{
+}
+
+/** @brief client-to-server streaming: a client sends tensors */
+Status
+ServiceImplProtobuf::SendTensors (ServerContext *context,
+    ServerReader<Tensors> *reader, Empty *reply)
+{
+  return _read_tensors (reader);
+}
+
+/** @brief server-to-client streaming: a client receives tensors */
+Status
+ServiceImplProtobuf::RecvTensors (ServerContext *context,
+    const Empty *request, ServerWriter<Tensors> *writer)
+{
+  return _write_tensors (writer);
+}
+
+template <typename T>
+Status ServiceImplProtobuf::_write_tensors (T writer)
+{
+  GstDataQueueItem *item;
+
+  /* until flushing */
+  while (gst_data_queue_pop (queue_, &item)) {
+    Tensors tensors;
+
+    _get_tensors_from_buffer (GST_BUFFER (item->object), tensors);
+    writer->Write (tensors);
+
+    g_free (item);
```

Comment:
for safe release
GDestroyNotify destroy = (item->destroy) ? item->destroy : g_free;
destroy (item);

Reply:
Thanks, updated.